### ISSUE TITLE
clrmd doesn't request pdbs for minidumps created without MiniDumpWithModuleHeaders flag

### ISSUE THREAD

**Issue by ezolenko** [microsoft/clrmd#63]
I'm parsing minidumps created as a last resort just before an unhandled exception escapes (in `AppDomain.CurrentDomain.UnhandledException`).

I'm finding that if I omit MiniDumpWithModuleHeaders when creating the dump, clrmd fails to request pdb files from symbol server when I'm going through exception stack frame and attempting to extract source info (exactly like in FileAndLineNumbers sample). Executable and dll modules are requested and downloaded as expected. 

In fact none of the enumerated modules (`DataTarget.EnumerateModules`) have pdb information.

When parsing dump created with MiniDumpWithModuleHeaders flag, pdbs are requested and loaded as expected (in most cases, I've seen similar problems I haven't narrowed down yet when exception originates in dynamically loaded assembly).

The problem with that flag is that it bloats 500 Kb dump into 25 Mb dump (slightly less convenient for mass collection). 

Visual Studio loads both kinds of minidumps and finds source info in both cases.

The problem likely related to DesktopModule loading PE info from minidump instead of directly reading binaries it downloaded from symbol server. (All PE headers end up with most of the directories, including  `DebugDirectory`, empty)

#### Dumped process arch
x64 (loads native dlls too)
#### .Net
4.6.2
#### OS 
win 10
#### Other minidump flags present in both cases
```
MiniDumpWithThreadInfo 
MiniDumpWithHandleData 
MiniDumpWithProcessThreadData 
MiniDumpWithUnloadedModules
MiniDumpWithTokenInformation
MiniDumpScanMemory
```

**Comment by leculver** [comment]
Sorry for the delay.

I haven't put this under a debugger directly yet, but it's my understanding that MiniDumpWithModuleHeaders does what it says...places module headers into the minidump.  For example, if you have a crash dump without MiniDumpWithModuleHeaders, is there any data at the beginning of a module?  In WinDbg:

    0:000> lmm clr
    start             end                 module name
    00007ffd`9abc0000 00007ffd`9b5a6000   clr        (deferred)        

If you dump memory at that address do you get the PE header or do you get nothing?  This is the PE header:

    0:000> dc 7ffd`9abc0000
    00007ffd`9abc0000  00905a4d 00000003 00000004 0000ffff  MZ..............

In order for any debugging tool to be able to locate PDBs, we need the first 1k of each module (or so), because that's where the PDB info is baked into the crash dump.  It's completely expected that ClrMD wouldn't be able to find the PDB in that case.

I can't speak to why that flag puts 25 megs into the crash dump...

**Comment by ezolenko** [comment]
If I understand the process correctly what happens is that clrmd attempts to read PE header from minidump and get pdb guid and age from there, while VS downloads the binary from symbol server first and then reads PE header from the complete downloaded binary before trying to find pdb. 

I think this workflow will fix this issue (and is more correct in principle):
- read module information from the dump
- find/download the module binary
- read PE header from actual file (without attempting to read PE header from minidump, because probably useless)
- find/download pdb based on that

I haven't looked if that's an easy change or if it requires a moderate refactoring though. I do see clrmd requesting and downloading PE binaries themselves from symbol server, so it has all the information necessary.

**Comment by leculver** [comment]
Ahh, using [https://msdn.microsoft.com/en-us/library/windows/desktop/ms680392(v=vs.85).aspx](MINIDUMP_MODULE).  I will take a look into this, but I've been considering some changes at the bottom level of clrmd.  It may have to wait until I can finish that work before I get back around to this.

**Comment by leculver** [comment]
I plan to replace the built-in minidump reader in ClrMD in a future release (currently thinking 1.2).  This would fix this problem.

**Comment by ryanmolden** [comment]
Not sure if relevant, but I also use MSDIA (presumably what CLRMD may use, or symsrv) in a private project and have found that dumps that don't contain module headers often WILL contain CodeView info (as part of the MINIDUMP_MODULE from the MINIDUMP_MODULE_LIST) and THAT can be used to load pdb symbols (MSDIA has a LoadDataFromCodeViewInfo).

**Comment by leculver** [comment]
This should now be fixed in ClrMD 2.0.
